File: crates/verify/ess-conformance/tests/fixtures/quoted-roundtrip.mjs
import assert from 'node:assert/strict'
import {readFileSync} from 'node:fs'
import {admitSuite, admitReplay} from '../../assets/coverage-admission.js'
const suite = await admitSuite(readFileSync(process.argv[2], 'utf8'))
const expected = JSON.parse(readFileSync(process.argv[3], 'utf8'))
const steps = suite.meaning['quoted.core/authored/operand'].steps
assert.equal(steps.length, expected.length)
for (const [index, step] of steps.entries()) {
  assert.deepEqual(step.expectation.predicate, ['compare', 'to', '==', ['literal', expected[index]]])
}
console.log(`${expected.length} actual Rust-emitted literals preserved by browser admission`)

for (const major of [5, 7]) {
  const original = JSON.parse(suite.original)
  original.provenance.suite_version = `ess-conformance/${major}`
  await assert.rejects(admitSuite(JSON.stringify(original)))
}
for (const badStep of [{step:'establish_entity'}, {step:'expect_response'}]) {
  const original = JSON.parse(suite.original)
  original.scenarios['quoted.core/authored/operand'].steps = [badStep]
  await assert.rejects(admitSuite(JSON.stringify(original)))
}
const original = JSON.parse(suite.original)
original.provenance.suite_version = 'ess-conformance/5'
original.scenarios['quoted.core/authored/operand'].steps = [{step:'expect_view', view:'quoted.core.Rows', expectation:{expect:'satisfies',predicate:{to:{eq:'"busy" status'}}}}]
await admitSuite(JSON.stringify(original))
console.log('Browser refuses old-version corrected operands and unsupported setup/response; compatible fallback stays suite/5')

const parentRef = {version:'ess-conformance/9', digest_profile:'sha256-json-bytes/1', digest:suite.digest}
const child = JSON.parse(suite.original)
child.coverage.selection.filter = {kind:'explicit', ids:Object.keys(child.scenarios), parent:parentRef}
const selected = await admitSuite(JSON.stringify(child))
const p = selected.document.provenance
const replay = {
  format:'ess-conformance-replay/1',
  model:{system:p.system, version:p.specification_version, spec_digest:p.spec_digest, contract_digest:p.contract_digest, entities:[], commands:[], views:[], actors:[], bindings:[]},
  suite:{...parentRef,digest:selected.digest},
  input:{format:'ess-conformance-input/1', suite_json:selected.original, parent_suites:[suite.original]}
}
await admitReplay(JSON.stringify(replay))
replay.suite.version = 'ess-conformance/5'
await assert.rejects(admitReplay(JSON.stringify(replay)))
console.log('Browser replay verifies actual suite/9 parent identity and refuses a false suite/5 reference')

File: crates/verify/ess-conformance/tests/fixtures/quoted-roundtrip.go
package essconform

import (
	"encoding/json"
	"os"
	"testing"
)

func TestQuotedRoundtrip(t *testing.T) {
	suite, err := admitRunInput(suiteJSON)
	if err != nil {
		t.Fatal(err)
	}
	var document struct {
		Scenarios map[string]struct {
			Steps []struct {
				Expectation struct{ Predicate json.RawMessage }
			}
		}
	}
	if err := json.Unmarshal([]byte(suite.original), &document); err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile("expected.json")
	if err != nil {
		t.Fatal(err)
	}
	var expected []string
	if err := json.Unmarshal(raw, &expected); err != nil {
		t.Fatal(err)
	}
	steps := document.Scenarios["quoted.core/authored/operand"].Steps
	if len(steps) != len(expected) {
		t.Fatal("missing cases")
	}
	for index, step := range steps {
		parsed, err := parsePredicate(step.Expectation.Predicate)
		if err != nil {
			t.Fatal(err)
		}
		if parsed.right.literal != expected[index] {
			t.Fatalf("case%d: %#v != %#v", index, parsed.right.literal, expected[index])
		}
		if parsed.evaluate(factSource{"to": expected[index]}) != truthTrue {
			t.Fatalf("case%d does not match", index)
		}
	}
	t.Logf("%d actual Rust-emitted literals preserved", len(expected))
}

func TestQuotedVersionAuthority(t *testing.T) {
	suite, err := admitRunInput(suiteJSON)
	if err != nil {
		t.Fatal(err)
	}
	var original map[string]any
	if err := json.Unmarshal([]byte(suite.original), &original); err != nil {
		t.Fatal(err)
	}
	if referenceFor(suite)["version"] != "ess-conformance/9" {
		t.Fatal("wrong parent version")
	}
	for _, major := range []string{"ess-conformance/5", "ess-conformance/7"} {
		original["provenance"].(map[string]any)["suite_version"] = major
		raw, _ := json.Marshal(original)
		if _, err := admitSuite(string(raw)); err == nil {
			t.Fatalf("accepted incompatible %s", major)
		}
	}
	for _, value := range []any{
		map[string]any{"to": map[string]any{"eq": "\"true\""}},
		map[string]any{"not": map[string]any{"to": map[string]any{"eq": "\"a.b\""}}},
		map[string]any{"forall": map[string]any{"in": "rows", "as": "row", "that": map[string]any{"to": map[string]any{"eq": "\"true\""}}}},
	} {
		if err := admitPredicateVersion(value, 7); err == nil {
			t.Fatalf("old admitted %#v", value)
		}
		if err := admitPredicateVersion(value, 8); err != nil {
			t.Fatal(err)
		}
	}
	for _, value := range []any{"to == \"a.b\"", map[string]any{"to": map[string]any{"eq": "\"busy\" status"}}} {
		if err := admitPredicateVersion(value, 4); err != nil {
			t.Fatal(err)
		}
	}
	t.Log("old/new version guard and actual coverage parent version checked")
}
